### PR TITLE
6979 pf script to c

### PR DESCRIPTION
**Pull request by javabalm**
|Related issue|
|---|
|#6979 #6978 |

## Description

This PR includes the translation of pf.sh and npf.sh scripts to C language

## Tests

- Compilation without warnings in every supported platform
  - [x] Linux
  - [x] FreeBSD
  - [x] NetBSD
  - [x] MAC OS X
- [x] Source installation
- [x] Package installation
- [x] Review logs syntax and correct language

### CODE REVIEW

File: src/active-response/firewalls/pf.c [hunk 1]
```diff
@@ -0,0 +1,164 @@
...
+                    memcpy(exec_cmd2, arg2, sizeof(exec_cmd2));
+                } else {
+                    char *arg1[7] = {PFCTL, "-t", PFCTL_TABLE, "-T", "delete", srcip, NULL};
+                    memcpy(exec_cmd1, arg1, sizeof(exec_cmd1));
+                }
+            } else {
+                memset(log_msg, '\0', LOGSIZE);
+                snprintf(log_msg, LOGSIZE - 1, "Table %s does not exist", PFCTL_TABLE);
+                write_debug_file(argv[0], log_msg);
+                return OS_INVALID;
```
Comment (TomasTurina) [inline src/active-response/firewalls/pf.c:106]: Old script returns `OS_SUCCESS`

File: src/active-response/firewalls/pf.c [hunk 2]
```diff
@@ -0,0 +1,164 @@
...
+            wpclose(wfd);
+        }
+
+    } else {
+        return OS_SUCCESS;
+    }
+
+    write_debug_file(argv[0], "Ended");
+
+    return 0;
```
Comment (TomasTurina) [inline src/active-response/firewalls/pf.c:145]: Use `OS_SUCCESS`

File: src/active-response/firewalls/pf.c [hunk 3]
```diff
@@ -0,0 +1,164 @@
...
+}
+
+int checking_if_its_configured(const char *path, const char *table) {
+    char command[1023];
```
Comment (TomasTurina) [inline src/active-response/firewalls/pf.c:149]: Use `COMMANDSIZE` here and in all the cases below. Use `COMMANDSIZE -1` in `snprintf`.

File: src/active-response/firewalls/pf.c [hunk 4]
```diff
@@ -0,0 +1,164 @@
...
+            write_debug_file(argv[0], log_msg);
+            return OS_SUCCESS;
+        }
+
+        // Executing it
+        if(strcmp(exec_cmd1[0], PFCTL) == 0) {
+            wfd_t *wfd = wpopenv(PFCTL, exec_cmd1, W_BIND_STDOUT);
+            if(!wfd) {
+                memset(log_msg, '\0', LOGSIZE);
+                snprintf(log_msg, LOGSIZE - 1, "Error executing %s : %s", PFCTL, strerror(errno));
```
Comment (TomasTurina) [inline src/active-response/firewalls/pf.c:121]: Use this log:
`"Error executing '%s' : %s"`

File: src/active-response/firewalls/pf.c [hunk 5]
```diff
@@ -0,0 +1,164 @@
...
+                return OS_INVALID;
+            }
+            wpclose(wfd);
+        }
+
+        if(strcmp(exec_cmd2[0], PFCTL) == 0) {
+            wfd_t *wfd = wpopenv(PFCTL, exec_cmd2, W_BIND_STDOUT);
+            if(!wfd) {
+                memset(log_msg, '\0', LOGSIZE);
+                snprintf(log_msg, LOGSIZE - 1, "Error executing %s : %s", PFCTL, strerror(errno));
```
Comment (TomasTurina) [inline src/active-response/firewalls/pf.c:132]: Use this log:
`"Error executing '%s' : %s"`

File: src/active-response/firewalls/npf.c [hunk 6]
```diff
@@ -0,0 +1,113 @@
...
+
+int main (int argc, char **argv) {
+    (void)argc;
+    char input[BUFFERSIZE];
+    char npf_active[COMMANDSIZE];
+    char npf_wazuh_ready[COMMANDSIZE];
+    char log_msg[LOGSIZE];
+    static char *srcip;
+    static char *action;
+    static char *iptables;
```
Comment (TomasTurina) [inline src/active-response/firewalls/npf.c:26]: Remove unused.

File: src/active-response/firewalls/npf.c [hunk 7]
```diff
@@ -0,0 +1,113 @@
...
+    if (system(npf_active) != 0) {
+        char log_msg[LOGSIZE] = "";
+        snprintf(log_msg, LOGSIZE -1, "Unable execute the command: '%s' ", npf_active);
+        write_debug_file(argv[0], log_msg);
+        return OS_INVALID;
+    }
+
+    memset(npf_wazuh_ready, '\0', COMMANDSIZE);
+    snprintf(npf_wazuh_ready, COMMANDSIZE -1,"%s show | grep -c 'table <wazuh_blacklist>'", NPFCTL);
+    if (system(npf_wazuh_ready) != 0) {
```
Comment (TomasTurina) [inline src/active-response/firewalls/npf.c:83]: Use `wpopen` instead of `system`, it is safer.

File: src/active-response/firewalls/npf.c [hunk 8]
```diff
@@ -0,0 +1,113 @@
...
+    cJSON_Delete(input_json);
+
+    if (access(NPFCTL, F_OK) < 0) {
+        write_debug_file(argv[0], "The NPFCTL is not accessible");
+        return OS_INVALID;
+    }
+
+    memset(npf_active, '\0', COMMANDSIZE);
+    snprintf(npf_active, COMMANDSIZE -1,"%s show | grep 'filtering:' | grep -c active", NPFCTL);
+    if (system(npf_active) != 0) {
```
Comment (TomasTurina) [inline src/active-response/firewalls/npf.c:74]: Use `wpopen` instead of `system`, it is safer.

File: src/active-response/firewalls/npf.c [hunk 9]
```diff
@@ -0,0 +1,113 @@
...
+        char log_msg[LOGSIZE] = "";
```
Comment (TomasTurina) [inline src/active-response/firewalls/npf.c:75]: Don't redefine log_msg, use:
`memset(log_msg, '\0', LOGSIZE);`

File: src/active-response/firewalls/npf.c [hunk 10]
```diff
@@ -0,0 +1,113 @@
...
+        snprintf(log_msg, LOGSIZE -1, "Unable execute the command: '%s' ", npf_active);
+        write_debug_file(argv[0], log_msg);
+        return OS_INVALID;
+    }
+
+    memset(npf_wazuh_ready, '\0', COMMANDSIZE);
+    snprintf(npf_wazuh_ready, COMMANDSIZE -1,"%s show | grep -c 'table <wazuh_blacklist>'", NPFCTL);
+    if (system(npf_wazuh_ready) != 0) {
+        char log_msg[LOGSIZE] = "";
```
Comment (TomasTurina) [inline src/active-response/firewalls/npf.c:84]: Don't redefine log_msg, use:
`memset(log_msg, '\0', LOGSIZE);`

File: src/active-response/firewalls/npf.c (hunk 8 above)
Comment (TomasTurina) [inline src/active-response/firewalls/npf.c:74]: Also, where do you check the value of `npf_active`?

File: src/active-response/firewalls/npf.c (hunk 7 above)
Comment (TomasTurina) [inline src/active-response/firewalls/npf.c:83]: Also, where do you check the value of `npf_wazuh_ready`?

File: src/active-response/firewalls/npf.c [hunk 11]
```diff
@@ -0,0 +1,113 @@
...
+        char *arg[6] = {NPFCTL, "table", "wazuh_blacklist", "del", srcip, NULL};
+        memcpy(exec_cmd, arg, sizeof(exec_cmd));
+
+    }
+
+    // Executing it
+    wfd_t *wfd = wpopenv(NPFCTL, exec_cmd, W_BIND_STDOUT);
+    if(!wfd) {
+        memset(log_msg, '\0', LOGSIZE);
+        snprintf(log_msg, LOGSIZE - 1, "Error executing %s : %s", NPFCTL, strerror(errno));
```
Comment (TomasTurina) [inline src/active-response/firewalls/npf.c:105]: Use this log:
`"Error executing '%s' : %s"`

File: src/active-response/firewalls/npf.c [hunk 12]
```diff
@@ -0,0 +1,113 @@
...
+        write_debug_file(argv[0], log_msg);
+        return OS_INVALID;
+    }
+    wpclose(wfd);
+
+    return 0;
```
Comment (TomasTurina) [inline src/active-response/firewalls/npf.c:111]: Use `OS_SUCCESS`.

**Review by TomasTurina — CHANGES_REQUESTED** [review]
Please, rebase your branch with `dev-ar-messages` before apply the necessary changes.

File: src/active-response/firewalls/pf.c [hunk 13]
```diff
@@ -0,0 +1,164 @@
+#include "shared.h"
+#include "external/cJSON/cJSON.h"
+#include "../active_responses.h"
```
[the remark this answers is not in the document]
Comment (javabalm) [reply src/active-response/firewalls/pf.c:3]: Done

File: src/active-response/firewalls/npf.c [hunk 14]
```diff
@@ -0,0 +1,113 @@
+#include "shared.h"
+#include "external/cJSON/cJSON.h"
+#include "../active_responses.h"
```
[the remark this answers is not in the document]
Comment (javabalm) [reply src/active-response/firewalls/npf.c:3]: Done

File: src/active-response/firewalls/npf.c [hunk 15]
```diff
@@ -0,0 +1,113 @@
...
+#include <stdlib.h>
+#include <stdbool.h>
+#include <errno.h>
+#include <unistd.h>
+#include <sys/stat.h>
+#include <sys/types.h>
+#include <sys/socket.h>
+#include <sys/utsname.h>
+
+#define LOG_FILE    "/logs/active-responses.log"
```
[the remark this answers is not in the document]
Comment (javabalm) [reply src/active-response/firewalls/npf.c:15]: Done

File: src/active-response/firewalls/npf.c [hunk 16]
```diff
@@ -0,0 +1,113 @@
...
+
+    // Get srcip
+    srcip = get_srcip_from_json(input_json);
+    if (!srcip) {
+        write_debug_file(argv[0], "Cannot read 'srcip' from data");
+        cJSON_Delete(input_json);
+        return OS_INVALID;
+    }
+
+    cJSON_Delete(input_json);
```
[the remark this answers is not in the document]
Comment (javabalm) [reply src/active-response/firewalls/npf.c:65]: Done

File: src/active-response/firewalls/pf.c [hunk 17]
```diff
@@ -0,0 +1,164 @@
...
+
+    // Get srcip
+    srcip = get_srcip_from_json(input_json);
+    if (!srcip) {
+        write_debug_file(argv[0], "Cannot read 'srcip' from data");
+        cJSON_Delete(input_json);
+        return OS_INVALID;
+    }
+
+    cJSON_Delete(input_json);
```
[the remark this answers is not in the document]
Comment (javabalm) [reply src/active-response/firewalls/pf.c:67]: Done

File: src/active-response/firewalls/pf.c [hunk 18]
```diff
@@ -0,0 +1,164 @@
...
+            write_debug_file(argv[0], log_msg);
+            return OS_SUCCESS;
+        }
+
+        char *exec_cmd1[7];
+        char *exec_cmd2[4];
+
+        // Checking if we have pf config file
+        if(access(PFCTL_RULES, F_OK) == 0) {
+            // Checking if ossec table is configured in pf.conf
```
[the remark this answers is not in the document]
Comment (javabalm) [reply src/active-response/firewalls/pf.c:90]: Done

File: src/active-response/firewalls/pf.c (hunk 1 above)
Comment (javabalm) [reply src/active-response/firewalls/pf.c:106]: Done

File: src/active-response/firewalls/pf.c (hunk 2 above)
Comment (javabalm) [reply src/active-response/firewalls/pf.c:145]: Done

File: src/active-response/firewalls/pf.c (hunk 3 above)
Comment (javabalm) [reply src/active-response/firewalls/pf.c:149]: Done

File: src/active-response/firewalls/pf.c (hunk 4 above)
Comment (javabalm) [reply src/active-response/firewalls/pf.c:121]: Done

File: src/active-response/firewalls/pf.c (hunk 5 above)
Comment (javabalm) [reply src/active-response/firewalls/pf.c:132]: Done

File: src/active-response/firewalls/npf.c (hunk 6 above)
Comment (javabalm) [reply src/active-response/firewalls/npf.c:26]: Done

File: src/active-response/firewalls/npf.c (hunk 12 above)
Comment (javabalm) [reply src/active-response/firewalls/npf.c:111]: Done

File: src/active-response/firewalls/npf.c (hunk 11 above)
Comment (javabalm) [reply src/active-response/firewalls/npf.c:105]: Done

File: src/active-response/firewalls/npf.c (hunk 7 above)
Comment (javabalm) [reply src/active-response/firewalls/npf.c:83]: Fixed

File: src/active-response/firewalls/npf.c (hunk 8 above)
Comment (javabalm) [reply src/active-response/firewalls/npf.c:74]: Fixed

File: src/active-response/firewalls/npf.c (hunk 10 above)
Comment (javabalm) [reply src/active-response/firewalls/npf.c:84]: Done

File: src/active-response/firewalls/npf.c (hunk 7 above)
Comment (javabalm) [reply src/active-response/firewalls/npf.c:83]: Fixed

File: src/active-response/firewalls/npf.c (hunk 8 above)
Comment (javabalm) [reply src/active-response/firewalls/npf.c:74]: Fixed

File: src/active-response/firewalls/npf.c (hunk 9 above)
Comment (javabalm) [reply src/active-response/firewalls/npf.c:75]: Done

**Review by TomasTurina — APPROVED** [review]
GJ!

**Pull request merged**